Stabilize `const_maybe_uninit_zeroed` and `const_mem_zeroed`

tgross35 opened the pull request:
Make `MaybeUninit::zeroed` and `mem::zeroed` const stable. Newly stable API:

```rust
// core::mem
pub const unsafe fn zeroed<T>() ->;

impl<T> MaybeUninit<T> {
    pub const fn zeroed() -> MaybeUninit<T>;
}
```

This relies on features based around `const_mut_refs`. Per @RalfJung, this should be OK since we do not leak any `&mut` to the user.

For this to be possible, intrinsics `assert_zero_valid` and `assert_mem_uninitialized_valid` were made const stable.

Tracking issue: #91850
Zulip discussion: https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/.60const_mut_refs.60.20dependents

r? libs-api
@rustbot label -T-libs +T-libs-api +A-const-eval 
cc @RalfJung  @oli-obk @rust-lang/wg-const-eval

tgross35 committed:
Stabilize `const_maybe_uninit_zeroed`

Make `MaybeUninit::zeroed` const stable. Newly stable API:

    // core::mem
    impl<T> MaybeUninit<T> {
        pub const fn zeroed() -> MaybeUninit<T>;
    }

Use of `const_mut_refs` should be acceptable since we do not leak the
mutability.

Tracking issue: #91850

tgross35 commented:
@rustbot label -T-libs

tgross35 commented:
cc implementer @lilasta

RalfJung commented:
LGTM, but the FCP should probably include t-lang since this makes the internal mutable referece/ptr support in our const-eval engine accessible from stable for the first time. I'm not at all concerned about doing so (that support is very mature), but it's worth a bit of extra scrutiny.

dtolnay commented:
@rfcbot fcp merge

scottmcm commented:
Should we make <https://doc.rust-lang.org/std/mem/fn.zeroed.html> `const` at the same time, since it's a simple wrapper?

But either way, no concerns from me about doing this.  Letting people have a `zeroinitializer` easily makes sense.

@rfcbot reviewed

RalfJung commented:
> Should we make https://doc.rust-lang.org/std/mem/fn.zeroed.html const at the same time, since it's a simple wrapper?

Sure, why not.

scottmcm commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -374,6 +374,20 @@ impl<T> MaybeUninit<T> {
+    ///
+    /// static PLUGIN_LIST: [PluginInfo; 2] = [
+    ///     PluginInfo { id: 1, action: Some(|x| x + 5) },
+    ///     unsafe { MaybeUninit::zeroed().assume_init() }
```
nit: could we find a better example? This could be `PluginInfo { id: 0, action: None }`, which is both shorter *and* doesn't need `unsafe`.

(And please put a `// SAFETY: …` comment on any `unsafe { … }` used in examples, to encourage good behaviour and explain why it's not UB.)

tgross35 replied:
Sure, any suggestions? I was struggling to come up with something better, null-terminated arrays for plugins are just the use case I always come across

Lokathor replied:
The entire problem with a good example here is that you use zeroed when the full thing would be a pain to write or even look at. So usually that means that you're talking about some type with 50 integer fields or something. But regardless of details it's "something so big it doesn't fit in a 2 line example".

dtolnay replied:
Are we able to use dev dependencies? Is there a `libc` use case to feature here?

Lokathor replied:
SafeZeroed would actually have to be an unsafe trait in this scenario.

Honestly I don't think we need example code on this particular function/method. It makes the all-zeroed value, that's already extremely simple I think.

tgross35 replied:
I really just wanted a sanity check test for the functionality, but agree it’s redundant. I’ll remove the example and change it to a unit test.

(Not in front of a computer now but I will do it tomorrow)

tgross35 replied:
Nevermind, I was able to do it now

@dtolnay`bot` ready 🙂

tgross35 committed:
Stabilize `const_mem_zeroed`

Make `core::mem::zeroed` const stable. Newly stable API:

    // core::mem
    pub const unsafe fn zeroed<T>() -> T;

This is stabilized with `const_maybe_uninit_zeroed` since it is a simple
wrapper.

In order to make this possible, intrinsics `assert_zero_valid` was made
const stable under `const_assert_type2`.
`assert_mem_uninitialized_valid` was also made const stable since it is
under the same gate.

tgross35 commented:
I updated this PR to do `mem::zeroed` as well. It depends on intrinsic `assert_zero_valid` becoming const, I assumed that isn't a problem either.

RalfJung commented:
Yeah, assert_inhabited, assert_zero_valid, assert_mem_uninitialized_valid are very similar and can just all be made stably const (since assert_inhabited already is stably const).

tgross35 commented:
Even easier 🙂 I added `assert_mem_uninitialized_valid`

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6348151607/job/17244268960) [(plain)](https://github.com/rust-lang/rust/commit/73a88fe2a1b9afceb4827db11541ae8c57bc0884/checks/17244268960/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=tgross35
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_20d6ff46-350f-42c6-ae1c-ae86a257a829
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=3f3c323efc17a425174286e962938ba3be53c97d
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_20d6ff46-350f-42c6-ae1c-ae86a257a829
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_20d6ff46-350f-42c6-ae1c-ae86a257a829
GITHUB_TRIGGERING_ACTOR=tgross35
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116218/merge
GITHUB_WORKFLOW_SHA=3f3c323efc17a425174286e962938ba3be53c97d
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
1 error[E0080]: evaluation of constant value failed
2   --> $DIR/assert-type-intrinsics.rs:12:9
3    |

16 LL |         intrinsics::assert_zero_valid::<&'static i32>();
17    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'aborted execution: attempted to zero-initialize type `&i32`, which is invalid', $DIR/assert-type-intrinsics.rs:20:9
- error: aborting due to 3 previous errors
+ error: aborting due to 3 previous errors; 1 warning emitted
20 
21 For more information about this error, try `rustc --explain E0080`.
---

error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/assert-type-intrinsics.rs:12:9
   |
LL |         MaybeUninit::<!>::uninit().assume_init();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'aborted execution: attempted to instantiate uninhabited type `!`', /checkout/tests/ui/consts/assert-type-intrinsics.rs:12:36
error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/assert-type-intrinsics.rs:16:9
   |
   |
LL |         intrinsics::assert_mem_uninitialized_valid::<&'static i32>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'aborted execution: attempted to leave type `&i32` uninitialized, which is invalid', /checkout/tests/ui/consts/assert-type-intrinsics.rs:16:9
error[E0080]: evaluation of constant value failed
##[error]  --> /checkout/tests/ui/consts/assert-type-intrinsics.rs:20:9
   |
   |
LL |         intrinsics::assert_zero_valid::<&'static i32>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'aborted execution: attempted to zero-initialize type `&i32`, which is invalid', /checkout/tests/ui/consts/assert-type-intrinsics.rs:20:9
error: aborting due to 3 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0080`.
------------------------------------------
```

</details>

WaffleLapkin commented:
> For this to be possible, intrinsics `assert_zero_valid` and `assert_mem_uninitialized_valid` were made const stable.

We should just use `#[rustc_allow_internal_unstable]` instead of stabilizing intrinsics.

nikomatsakis commented:
Ralf can you elaborate on what this means:

> LGTM, but the FCP should probably include t-lang since this makes the internal mutable referece/ptr support in our const-eval engine accessible from stable for the first time.

What is "internal mutable reference support"? @scottmcm suggested that it meant "a mutable reference appears in the implementation, but it is not leaked out and does not permit `&mut` to be used more broadly"?

I have to admit that I don't have a solid mental model of what the exact invariants are and the edges to stay away from when it comes to const eval. This stabilization certainly *feels* harmless, but I'd like to understand better. What happens if you do `MaybeUninit::<&mut 32>::zeroed`?

tgross35 commented:
> > For this to be possible, intrinsics `assert_zero_valid` and `assert_mem_uninitialized_valid` were made const stable.
> 
> We should just use `#[rustc_allow_internal_unstable]` instead of stabilizing intrinsics.

@WaffleLapkin That just comes from Ralf's suggestion https://github.com/rust-lang/rust/pull/116218#issuecomment-1740395402, since the similar `assert_inhabited` is already const stable

RalfJung commented:
The intrinsic isn't stable, it's const-stable. That's what we generally do for intrinsics that can be (transitively) called in const. This does *not* make the intrinsic directly callable on stable (neither in const nor outside const).

tgross35 commented:
cc @rust-lang/libs-api since discussion is active

RalfJung commented:
We don't stabilize intrinsics so I cannot see the footgun here.

nikomatsakis commented:
@rfcbot reviewed

I don't 100% understand the rules that are preventing leakage here, and would like to better, but I see no reason to block stabilization. I am quite confident we will always want to support `&mut` usage encapsulated wihtin const fn bodies.

tgross35 commented:
I can’t speak definitively but I think that the main issues with `&mut` relate to interior mutability, which isn’t needed here.

As an extra safeguard for this specific feature, we should be able to have a zero initializer that doesn’t rely on const `&mut` if that ever goes away for whatever reason.

RalfJung commented:
Interior mutability is not related. But indeed the end to end behavior of "zeroed" is a pure function without mutating any outside state - mutability is encapsulated. That's why stabilizing it is fine even if &mut still has kinks that need to be worked out.

dtolnay reviewed (commented):
Thanks!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 5e5f3341e3749f18a6d6fbb2228a9d3b42c0b280 has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay 5e5f3341e3749f18a6d6fbb2228a9d3b42c0b280 -->
<!-- homu: {"type":"Approved","sha":"5e5f3341e3749f18a6d6fbb2228a9d3b42c0b280","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 5e5f3341e3749f18a6d6fbb2228a9d3b42c0b280 with merge da1e0d1d75bc029adfa30c596b526bc17436453e...
<!-- homu: {"type":"BuildStarted","head_sha":"5e5f3341e3749f18a6d6fbb2228a9d3b42c0b280","merge_sha":"da1e0d1d75bc029adfa30c596b526bc17436453e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6759542716/job/18373300461)
Approved by: dtolnay
Pushing da1e0d1d75bc029adfa30c596b526bc17436453e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"dtolnay","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6759542716/job/18373300461"},"merge_sha":"da1e0d1d75bc029adfa30c596b526bc17436453e"} -->

rust-timer commented:
Finished benchmarking commit (da1e0d1d75bc029adfa30c596b526bc17436453e): [comparison URL](https://perf.rust-lang.org/compare.html?start=f64d028196e06ca523ac5de442dc4772cfa58b47&end=da1e0d1d75bc029adfa30c596b526bc17436453e&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=f64d028196e06ca523ac5de442dc4772cfa58b47&end=da1e0d1d75bc029adfa30c596b526bc17436453e&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f64d028196e06ca523ac5de442dc4772cfa58b47&end=da1e0d1d75bc029adfa30c596b526bc17436453e&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.4%, 0.7%]   | 4     |
| Regressions ❌ <br /> (secondary)  | 3.3%  | [3.2%, 3.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.8%, -0.4%] | 5     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.1% | [-0.8%, 0.7%]  | 9     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f64d028196e06ca523ac5de442dc4772cfa58b47&end=da1e0d1d75bc029adfa30c596b526bc17436453e&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.5%, 0.7%]   | 4     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-0.7%, -0.7%] | 1     |
| All ❌✅ (primary)                 | 0.6%  | [0.5%, 0.7%]   | 4     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f64d028196e06ca523ac5de442dc4772cfa58b47&end=da1e0d1d75bc029adfa30c596b526bc17436453e&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 638.639s -> 638.563s (-0.01%)
**Artifact size**: 304.53 MiB -> 304.34 MiB (-0.06%)

the8472 commented on library/core/tests/mem.rs:
```diff
@@ -565,3 +565,24 @@ fn offset_of_addr() {
 }
+
+#[test]
+fn const_maybe_uninit_zeroed() {
+    // Sanity check for `MaybeUninit::zeroed` in a realistic const situation (plugin array term)
+    #[repr(C)]
+    struct Foo {
+        a: Option<&'static str>,
+        b: Bar,
+        c: f32,
+        d: *const u8,
+    }
+    #[repr(C)]
+    struct Bar(usize);
+    struct FooPtr(*const Foo);
+    unsafe impl Sync for FooPtr {}
+
+    static UNINIT: FooPtr = FooPtr([unsafe { MaybeUninit::zeroed().assume_init() }].as_ptr());
```
I'm not sure what motivated this test but there actually isn't any guarantee that all-zero bytes are a valid value for `Option<&'static str>` since that's a fat pointer.

[NPO is only guaranteed](https://doc.rust-lang.org/nightly/std/option/index.html#representation) for thin pointers

RalfJung replied:
Yeah but just because there is no guarantee doesn't mean we should have no test.

Or put differently, library and ui tests are a rustc implementation detail and not source for stable guarantees. Only docs are stable guarantees.

Lokathor replied:
Also we *should* make that a docs guarantee for &str, as a separate future PR

the8472 replied:
That'd conflict with niches-for-metadata that I'm working on. The length field will have a larger niche which means it becomes preferred by the layout algorithm and Option::None is no longer zero but some other value.
That's how I found this.

> Or put differently, library and ui tests are a rustc implementation detail and not source for stable guarantees. 

I mean that this looks like it has been inspired by some real code and that that code is suspect.

Lokathor replied:
That sounds extremely unfortunate and that a special case should be put back in or something. But we can go into that on zulip or whatever PR you open.

tgross35 replied:
This test was just a sanity check that it does indeed do the right thing. I suppose `&str` might not have been the best optional reference, but that specific choice wasn't motivated by any real code.

Otoh, it should still zero everything even if invalid, even if Miri one day starts flagging this test.

the8472 replied:
```
error[E0080]: it is undefined behavior to use this value
   --> library/core/tests/mem.rs:789:5
    |
789 |     static UNINIT: FooPtr = FooPtr([unsafe { MaybeUninit::zeroed().assume_init() }].as_ptr());
    |     ^^^^^^^^^^^^^^^^^^^^^ constructing invalid value at .<deref>[0].a.<enum-variant(Some)>.0: encountered a null reference
    |
    = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rustc repository if you believe it should not be considered undefined behavior.
    = note: the raw bytes of the constant (size: 8, align: 8) {
                ╾───alloc7286<imm>────╼                         │ ╾──────╼
            }
```

tgross35 replied:
> Otoh, it should still zero everything even if invalid, even if Miri one day starts flagging this test.

...or the compiler itself I suppose. Feel free to change or `#[allow(...)]` this test as needed, as long as there isn't any risk of something like `MaybeUninit::zeroed()`'s output later being fixed up with a non-zero niche in the pointer metadata (seems unlikely)

RalfJung replied:
Even if we never guaranteed NPO for wide pointers, changing that seems very risky and I expect it will break some unsafe code out there. Most people won't expect such a discontinuity between thin and wide ptrs.
